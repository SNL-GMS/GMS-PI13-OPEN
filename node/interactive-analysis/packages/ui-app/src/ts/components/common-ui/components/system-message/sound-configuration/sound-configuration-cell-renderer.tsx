import { ColumnDefinition, DropDown, TableCellRenderer } from '@gms/ui-core-components';
import cloneDeep from 'lodash/cloneDeep';
import * as React from 'react';
import { userPreferences } from '~components/common-ui/config/user-preferences';
import { SoundConfigurationRendererParams, SoundConfigurationRow } from './types';

export type SoundConfigurationColumnDefinition = ColumnDefinition<
  SoundConfigurationRow,
  {},
  string | number,
  {},
  {}
>;

const defaultColumnWidthPx = 200;

const getColumnPosition = (props: SoundConfigurationRendererParams): number | 'first' | 'last' => {
  const index = props.columnApi
    .getAllDisplayedColumns()
    .findIndex(c => c.getColId() === props.colDef.colId);
  return index === 0
    ? 'first'
    : index === props.columnApi.getAllDisplayedColumns().length - 1
    ? 'last'
    : index;
};

export const NotificationsStatusCellRenderer: React.FunctionComponent<SoundConfigurationRendererParams> = props =>
  props.data?.hasNotificationStatusError ? (
    <TableCellRenderer
      className={`sound-configuration--notification-status-no-file-error`}
      data-col-position={getColumnPosition(props)}
      value={'!'}
      tooltipMsg={userPreferences.configuredAudibleNotificationFileNotFound(
        props.data.sound.selectedSound
      )}
      shouldCenterText={true}
    >
      {props.children}
    </TableCellRenderer>
  ) : null;

export const SoundConfigurationCellRenderer: React.FunctionComponent<SoundConfigurationRendererParams> = props => (
  <TableCellRenderer
    data-col-position={getColumnPosition(props)}
    value={props.valueFormatted ?? props.value}
  />
);

export const SoundConfigurationDropdownRenderer: React.FunctionComponent<SoundConfigurationRendererParams> = props => {
  if (!props.data || !props.data.sound) return null;

  const availableSounds = cloneDeep(props.data.sound.availableSounds);

  // if the configured sound happens to no longer be listed in the available sounds; add it
  // this will allow the selection to be correct and for an error to be displayed
  if (
    props.data.sound.selectedSound !== 'None' &&
    !availableSounds[props.data.sound.selectedSound]
  ) {
    availableSounds[props.data.sound.selectedSound] = props.data.sound.selectedSound;
  }
  return (
    <DropDown
      className={'sound-configuration__dropdown'}
      dropDownItems={availableSounds}
      value={props.data.sound.selectedSound}
      onMaybeValue={v => props.data.sound.onSelect(v)}
      title={'Select a sound'}
    />
  );
};

const headerCellBlockClass = 'soh-header-cell';

export const defaultColumnDefinition: SoundConfigurationColumnDefinition = {
  headerClass: `${headerCellBlockClass} ${headerCellBlockClass}--neutral`,
  width: defaultColumnWidthPx,
  sortable: true,
  filter: true,
  disableStaticMarkupForHeaderComponentFramework: true,
  disableStaticMarkupForCellRendererFramework: true,
  cellRendererFramework: SoundConfigurationCellRenderer
};
