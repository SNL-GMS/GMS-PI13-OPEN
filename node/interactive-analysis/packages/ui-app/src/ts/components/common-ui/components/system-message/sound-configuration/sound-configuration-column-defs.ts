import { ColumnDefinition } from '@gms/ui-core-components';
import {
  NotificationsStatusCellRenderer,
  SoundConfigurationDropdownRenderer
} from './sound-configuration-cell-renderer';
import { SoundConfigurationRow } from './types';

export const columnDefs: ColumnDefinition<SoundConfigurationRow, {}, {}, {}, {}>[] = [
  {
    headerName: '',
    field: 'hasNotificationStatusError',
    width: 40,
    minWidth: 40,
    resizable: false,
    suppressMovable: true,
    cellRendererFramework: NotificationsStatusCellRenderer,
    suppressFilter: true,
    suppressSorting: true
  },
  {
    headerName: 'Sound',
    field: 'sound',
    cellRendererFramework: SoundConfigurationDropdownRenderer
  },
  {
    headerName: 'Category',
    field: 'category'
  },
  {
    headerName: 'Subcategory',
    field: 'subcategory'
  },
  {
    headerName: 'Severity',
    field: 'severity'
  },
  {
    headerName: 'Message',
    field: 'message',
    flex: 1,
    minWidth: 300
  }
];
