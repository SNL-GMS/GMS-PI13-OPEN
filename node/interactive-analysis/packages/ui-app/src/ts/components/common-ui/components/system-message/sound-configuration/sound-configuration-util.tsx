import { UserProfileTypes } from '@gms/common-graphql';
import { AVAILABLE_SOUND_FILES } from '@gms/common-util';
import { UILogger } from '@gms/ui-apollo';
import { Toaster } from '@gms/ui-util';
import Axios from 'axios';
import includes from 'lodash/includes';
import sortBy from 'lodash/sortBy';
import uniq from 'lodash/uniq';
import { userPreferences } from '~components/common-ui/config/user-preferences';
import { ALL_CATEGORIES, ALL_SEVERITIES, ALL_SUBCATEGORIES, SoundConfigurationRow } from './types';

const BASE_SOUNDS_PATH = userPreferences.baseSoundsPath;

const toaster = new Toaster();

/**
 * Determine which sound files are missing, if any.
 * @returns the missing sound files; empty array if no files are missing
 */
export const getMissingSounds = async (filenames: string[]): Promise<string[]> =>
  Promise.all<string>(
    sortBy<string>(uniq<string>(filenames)).map(async soundFileName =>
      Axios.get(`${BASE_SOUNDS_PATH}${String(soundFileName)}`)
        .then(response => undefined)
        .catch(error => soundFileName)
    )
  ).then(response => response.filter(r => r !== undefined));

/**
 * Validate the available sound files. Toast an error message for
 * each sound file that is not loaded properly (not found).
 */
export const validateAvailableSounds = (): void => {
  getMissingSounds(AVAILABLE_SOUND_FILES)
    .then(missingSoundFiles => {
      missingSoundFiles.forEach(soundFileName => {
        toaster.toastError(userPreferences.soundFileNotFound(soundFileName));
        UILogger.Instance().error(
          userPreferences.soundFileNotFound(`${BASE_SOUNDS_PATH}${soundFileName}`)
        );
      });
    })
    .catch(error => UILogger.Instance().error(error));
};

/**
 * Validate the configured audible notifications against the available sounds.
 * Toast an error message for each sound file that is configured for an audible notifications but not available.
 */
export const validateConfiguredAudibleNotifications = (
  audibleNotifications: UserProfileTypes.AudibleNotification[]
): void => {
  const configuredSoundFiles = sortBy<string>(
    uniq<string>(audibleNotifications?.map(notification => notification.fileName))
  );

  getMissingSounds(configuredSoundFiles)
    .then(missingSoundFiles => {
      const soundFilesInError = configuredSoundFiles.filter(sound =>
        includes(missingSoundFiles, sound)
      );
      soundFilesInError.forEach(soundFileName => {
        toaster.toastError(
          userPreferences.configuredAudibleNotificationFileNotFound(soundFileName)
        );
        UILogger.Instance().error(
          userPreferences.configuredAudibleNotificationFileNotFound(
            `${BASE_SOUNDS_PATH}${soundFileName}`
          )
        );
      });
    })
    .catch(error => UILogger.Instance().error(error));
};

/**
 * @returns the filenames of the sound files that are in the /sounds directory
 * Relative path only. For example /sounds/example.mp3 would be returned as "example.mp3"
 */
export const getAvailableSounds = () => {
  const soundFileNames: string[] = AVAILABLE_SOUND_FILES;
  const obj: { [key: string]: string } = {};
  let soundsMap = soundFileNames.reduce((prev, curr) => {
    prev[curr] = curr;
    return prev;
  }, obj);
  soundsMap = { default: 'None', ...soundsMap };
  return soundsMap;
};

/**
 * Filter rows based on severity, category, and subcategory
 * @param selectedSeverity - selected severity
 * @param rows - The rows to be filtered
 * @param selectedCategory - Selected category
 * @param selectedSubcategory - Selected subcategory
 *
 * @returns - Filtered sound configuration rows
 */
export const filterRowsByType = (
  selectedSeverity: string,
  rows: SoundConfigurationRow[],
  selectedCategory: string,
  selectedSubcategory: string
) => {
  let newRows = [...rows];

  if (selectedSeverity !== ALL_SEVERITIES) {
    newRows = newRows.filter(row => row.severity === selectedSeverity);
  }
  if (selectedCategory !== ALL_CATEGORIES) {
    newRows = newRows.filter(row => row.category === selectedCategory);
  }
  if (selectedSubcategory !== ALL_SUBCATEGORIES) {
    newRows = newRows.filter(row => row.subcategory === selectedSubcategory);
  }
  return newRows;
};
