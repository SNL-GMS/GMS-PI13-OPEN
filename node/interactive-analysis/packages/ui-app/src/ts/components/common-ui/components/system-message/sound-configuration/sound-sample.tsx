import { UILogger } from '@gms/ui-apollo';
import React, { createRef, useEffect, useState } from 'react';

interface SoundSampleProps {
  soundToPlay: string;
}

export const SoundSample: React.FunctionComponent<SoundSampleProps> = (props: SoundSampleProps) => {
  const ref = createRef<HTMLAudioElement>();
  const filename: string = props.soundToPlay?.split('/').slice(-1)[0];

  const [play, setPlay] = useState(false);

  useEffect(() => {
    if (play && props.soundToPlay && ref.current && filename !== 'None') {
      ref.current.play().catch(e => {
        UILogger.Instance().error(`Error playing sound "${props.soundToPlay}": ${e}`);
      });
    }
    setPlay(true);
  }, [props.soundToPlay]);

  if (filename === 'None') {
    return null;
  }
  return <audio ref={ref} src={props.soundToPlay} autoPlay={false} />;
};
