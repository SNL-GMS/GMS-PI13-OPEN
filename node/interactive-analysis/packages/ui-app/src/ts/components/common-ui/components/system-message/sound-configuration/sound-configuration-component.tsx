import { Button, Card, Dialog } from '@blueprintjs/core';
import { SystemMessageTypes, UserProfileTypes } from '@gms/common-graphql';
import { UILogger } from '@gms/ui-apollo';
import { InfiniteTable, useDatasource } from '@gms/ui-core-components';
import classNames from 'classnames';
import includes from 'lodash/includes';
import React, { useState } from 'react';
import { userPreferences } from '~components/common-ui/config/user-preferences';
import {
  getHeaderHeight,
  getRowHeightWithBorder
} from '~components/data-acquisition-ui/shared/table/utils';
import { AudibleNotificationContext } from '../audio/audible-notification-context';
import { defaultColumnDefinition } from './sound-configuration-cell-renderer';
import { columnDefs } from './sound-configuration-column-defs';
import { SoundConfigurationToolbar } from './sound-configuration-toolbar';
import { filterRowsByType, getAvailableSounds, getMissingSounds } from './sound-configuration-util';
import { SoundSample } from './sound-sample';
import {
  ALL_CATEGORIES,
  ALL_SEVERITIES,
  ALL_SUBCATEGORIES,
  FILTER_TYPE,
  SelectedOptions as SelectedFilterOptions,
  SoundConfigurationRow
} from './types';

interface SoundConfigurationProps {
  isVisible: boolean;
  systemMessageDefinitions: SystemMessageTypes.SystemMessageDefinition[];
  onToggle(): void;
}

/**
 * Custom hook to manage which rows are filtered and which are not.
 * @param rows the rows to filter
 * @returns
 * an the filtered rows;
 * a callback function to call when a filter is changed;
 * and a SelectedFilterOptions object containing the state of each filter
 */
const useRowFilters = (
  rows: SoundConfigurationRow[]
): [SoundConfigurationRow[], (type: FILTER_TYPE, value: any) => void, SelectedFilterOptions] => {
  const [selectedSeverity, setSelectedSeverity] = useState(ALL_SEVERITIES);
  const [selectedCategory, setSelectedCategory] = useState(ALL_CATEGORIES);
  const [selectedSubcategory, setSelectedSubcategory] = useState(ALL_SUBCATEGORIES);

  const onFilterChanged = (type: FILTER_TYPE, value): void => {
    switch (type) {
      case FILTER_TYPE.SEVERITY:
        setSelectedSeverity(value);
        break;
      case FILTER_TYPE.CATEGORY:
        setSelectedCategory(value);
        break;
      case FILTER_TYPE.SUBCATEGORY:
        setSelectedSubcategory(value);
        break;
      default:
        // This should never happen...
        // tslint:disable-next-line: no-console
        console.error('Unknown SystemMessage category');
    }
  };

  const filteredRows = filterRowsByType(
    selectedSeverity,
    rows,
    selectedCategory,
    selectedSubcategory
  );
  return [
    filteredRows,
    onFilterChanged,
    {
      selectedSeverity,
      selectedCategory,
      selectedSubcategory
    }
  ];
};

/**
 * creates a list of sound configuration rows.
 * @param systemMessageDefinitions the defined system messages from which to generate rows
 * @param onSelect called when a selection is made.
 */
const useSoundConfigurationRows = (
  systemMessageDefinitions: SystemMessageTypes.SystemMessageDefinition[],
  tableRef: React.MutableRefObject<InfiniteTable<SoundConfigurationRow, {}>>,
  onSelect?: (fileName: string) => void
): SoundConfigurationRow[] => {
  const context = React.useContext(AudibleNotificationContext);

  const [missingSoundFiles, setMissingSoundFiles] = useState<string[]>(null);

  const configuredAudibleNotifications = context.audibleNotifications?.map(a => a.fileName);

  React.useEffect(() => {
    getMissingSounds(configuredAudibleNotifications)
      .then(response => {
        setMissingSoundFiles(response);
      })
      .catch(error => UILogger.Instance().error(error));
  }, [context]);

  const availableSounds = getAvailableSounds();

  const getSelectedSound = (
    systemMessageDefinition: SystemMessageTypes.SystemMessageDefinition
  ): string =>
    context.audibleNotifications?.find(
      notification =>
        SystemMessageTypes.SystemMessageType[notification.notificationType] ===
        SystemMessageTypes.SystemMessageType[systemMessageDefinition.systemMessageType]
    )?.fileName ?? 'None';

  const rows: SoundConfigurationRow[] = systemMessageDefinitions.map(
    (systemMessageDefinition, idx) => ({
      id: `${idx}`,
      sound: {
        availableSounds,
        selectedSound: getSelectedSound(systemMessageDefinition),
        onSelect: selectedFileName => {
          const notification: UserProfileTypes.AudibleNotification = {
            fileName: selectedFileName === 'None' ? '' : selectedFileName,
            notificationType:
              SystemMessageTypes.SystemMessageType[systemMessageDefinition.systemMessageType]
          };

          context
            .setAudibleNotifications([notification])
            .then(result => {
              onSelect(selectedFileName);
              const tableApi = tableRef.current.getTableApi();
              tableApi.refreshInfiniteCache();
            })
            .catch(err => {
              UILogger.Instance().error(err);
            });
        }
      },
      hasNotificationStatusError:
        missingSoundFiles &&
        missingSoundFiles.length > 0 &&
        includes(missingSoundFiles, getSelectedSound(systemMessageDefinition)),
      category: systemMessageDefinition.systemMessageCategory,
      subcategory: systemMessageDefinition.systemMessageSubCategory,
      severity: systemMessageDefinition.systemMessageSeverity,
      message: systemMessageDefinition.template
    })
  );

  return rows;
};

/**
 * The actual component that creates a dialog, chooses and plays sounds.
 * Calls on the context.setAudibleNotifications to update the user profile as
 * soon as user selects a sound.
 */
const SoundConfigurationComponent: React.FunctionComponent<SoundConfigurationProps> = ({
  isVisible,
  onToggle,
  systemMessageDefinitions
}) => {
  const [sampleSound, setSampleSound] = useState<string>(null);
  const tableRef = React.useRef<InfiniteTable<SoundConfigurationRow, {}>>(null);
  const setSampleSoundWithPath = (s: string) =>
    setSampleSound(`${userPreferences.baseSoundsPath}${s}`);
  const rows = useSoundConfigurationRows(
    systemMessageDefinitions,
    tableRef,
    setSampleSoundWithPath
  );
  const [filteredRows, onFilterChanged, selectedOptions] = useRowFilters(rows);
  const datasource = useDatasource(filteredRows);

  return (
    <Dialog
      onClose={onToggle}
      isOpen={isVisible}
      autoFocus={true}
      hasBackdrop={true}
      canEscapeKeyClose={true}
      title={'Sound Configuration'}
      style={{ width: 'auto' }}
      canOutsideClickClose={true}
    >
      <div className={classNames('sound-configuration ag-theme-dark')}>
        <SoundSample soundToPlay={sampleSound} />
        <Card interactive={false}>
          <SoundConfigurationToolbar
            systemMessageDefinitions={systemMessageDefinitions}
            selectedOptions={selectedOptions}
            onChanged={onFilterChanged}
          />
          <div className={'sound-configuration__container'}>
            <InfiniteTable<SoundConfigurationRow, {}>
              ref={ref => (tableRef.current = ref)}
              id="table-sound-configuration"
              key="table-sound-configuration"
              datasource={datasource}
              context={{}}
              defaultColDef={defaultColumnDefinition}
              columnDefs={columnDefs}
              rowData={filteredRows}
              rowHeight={getRowHeightWithBorder()}
              headerHeight={getHeaderHeight()}
              getRowNodeId={node => node.id}
              deltaRowDataMode={true}
              rowDeselection={true}
              overlayNoRowsTemplate="No sounds to display"
              suppressContextMenu={true}
            />
          </div>
          <div className={'sound-configuration__footer'}>
            <Button className={'sound-configuration__close-button'} onClick={onToggle}>
              Close
            </Button>
          </div>
        </Card>
      </div>
    </Dialog>
  );
};

/**
 * Dialog sound configuration component. The purpose of this component is to give
 * the user the ability to select a sound from a predefined list for specific
 * events that happen.
 */
export const SoundConfiguration = React.memo(SoundConfigurationComponent);
