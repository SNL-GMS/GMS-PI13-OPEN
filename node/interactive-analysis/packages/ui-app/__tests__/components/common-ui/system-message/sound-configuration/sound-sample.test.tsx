import * as React from 'react';

import { SoundSample } from '../../../../../src/ts/components/common-ui/components/system-message/sound-configuration/sound-sample';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

describe('Sound Sample Component', () => {
  const soundSample = Enzyme.mount(<SoundSample soundToPlay={'test.mp3'} />);

  it('should be defined', () => {
    expect(SoundSample).toBeDefined();
    expect(soundSample).toBeDefined();
  });

  it('matches snapshot', () => {
    soundSample.update();
    expect(soundSample).toMatchSnapshot();
  });

  const noneSound = Enzyme.mount(<SoundSample soundToPlay={'None'} />);

  it("validates 'None' sound", () => {
    noneSound.update();
    expect(noneSound).toMatchSnapshot();
  });
});
