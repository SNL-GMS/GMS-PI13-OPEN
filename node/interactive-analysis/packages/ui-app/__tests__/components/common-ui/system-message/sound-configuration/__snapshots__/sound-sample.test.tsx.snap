// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sound Sample Component matches snapshot 1`] = `
<Component
  soundToPlay="test.mp3"
>
  <audio
    autoPlay={false}
    src="test.mp3"
  />
</Component>
`;

exports[`Sound Sample Component validates 'None' sound 1`] = `
<Component
  soundToPlay="None"
/>
`;
